File: 801-is-graph-bipartite/is-graph-bipartite.cpp
class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        int n=graph.size();
        vector<int>V(n,-1);
        queue<int>Q;
        for(int i=0;i<n;i++){
            int start=i;
            if(V[start]!=-1) continue;
            V[start]=1;
            Q.push(start);
            while(!Q.empty()){
                int parent=Q.front();
                int next_color=!V[parent];
                Q.pop();
                for(auto child:graph[parent]){
                    if(V[child]==-1){
                        V[child]=next_color;
                        Q.push(child);
                    }
                    else{
                        if(V[child]!=next_color) return false;
                    }
                }
            }
        }
        return true;
    }
};

File: 801-is-graph-bipartite/is-graph-bipartite.py
from collections import deque
from typing import List

class Solution:
    def isBipartite(self, graph: List[List[int]]) -> bool:
        n = len(graph)
        color = [-1] * n 
        
        for start in range(n):
            if color[start] == -1: 
                queue = deque([start])
                color[start] = 0
                
                while queue:
                    node = queue.popleft()
                    for nei in graph[node]:
                        if color[nei] == -1:
                            color[nei] = 1 - color[node]
                            queue.append(nei)
                        elif color[nei] == color[node]:
                            return False
        return True